tool/tsh/common/mcp_test.go

```diff
@@ -0,0 +1,300 @@
+	alice.SetDatabaseNames([]string{"postgres"})
+	alice.SetRoles([]string{"access"})
+
+	authProcess := testserver.MakeTestServer(
+		t,
+		testserver.WithBootstrap(connector, alice),
+		testserver.WithConfig(func(cfg *servicecfg.Config) {
+			cfg.Auth.NetworkingConfig.SetProxyListenerMode(types.ProxyListenerMode_Multiplex)
+			cfg.Databases.Enabled = true
+			cfg.Databases.Databases = []servicecfg.Database{
+				{
+					Name:     "postgres-local",
+					Protocol: defaults.ProtocolPostgres,
+					URI:      "external-pg:5432",
+				},
+				{
+					Name:     "mysql-local",
+					Protocol: defaults.ProtocolMySQL,
+					URI:      "external-mysql:3306",
+				},
+			}
+		}),
+	)
+
+	authServer := authProcess.GetAuthServer()
+	require.NotNil(t, authServer)
+
+	proxyAddr, err := authProcess.ProxyWebAddr()
+	require.NoError(t, err)
+
+	err = Run(t.Context(), []string{
+		"login", "--insecure", "--debug", "--proxy", proxyAddr.String(),
+	}, setHomePath(tmpHomePath), setMockSSOLogin(authServer, alice, connector.GetName()))
+	require.NoError(t, err)
+
+	stdin, writer := io.Pipe()
+	reader, stdout := io.Pipe()
+
+	ctx, cancel := context.WithCancel(t.Context())
+	defer cancel()
+
+	executionCh := make(chan error)
+	go func() {
+		executionCh <- Run(ctx, []string{
+			"mcp", "db", "--insecure", "--db-user=postgres", "--db-name=postgres",
+		}, setHomePath(tmpHomePath), func(c *CLIConf) error {
+			c.overrideStdin = stdin
+			c.OverrideStdout = stdout
+			// MCP server logs are going to be discarded.
+			c.overrideStderr = io.Discard
+			c.databaseMCPRegistryOverride = map[string]dbmcp.NewServerFunc{
+				defaults.ProtocolPostgres: func(ctx context.Context, nsc *dbmcp.NewServerConfig) (dbmcp.Server, error) {
+					return &testDatabaseMCP{}, nil
+				},
+			}
+			return nil
+		})
+	}()
+
+	clt := mcpclient.NewClient(NewStdio(writer, bufio.NewReader(reader)))
+	require.NoError(t, clt.Start(t.Context()))
+
+	req := mcp.InitializeRequest{}
+	req.Params.ProtocolVersion = mcp.LATEST_PROTOCOL_VERSION
+	req.Params.ClientInfo = mcp.Implementation{
+		Name:    "test-client",
+		Version: "1.0.0",
+	}
+
+	require.EventuallyWithT(t, func(collect *assert.CollectT) {
+
+		_, err = clt.Initialize(t.Context(), req)
+		require.NoError(collect, err)
+		require.NoError(collect, clt.Ping(t.Context()))
+	}, time.Second, 100*time.Millisecond)
+
+	// Stop the MCP server command and wait until it is finshed.
+	cancel()
+	select {
+	case err := <-executionCh:
+		require.Error(t, err)
+		require.ErrorIs(t, err, context.Canceled)
+	case <-time.After(10 * time.Second):
+		require.Fail(t, "expected the execution to be completed")
+	}
+}
+
+// testDatabaseMCP is a noop database MCP server.
+type testDatabaseMCP struct{}
+
+func (s *testDatabaseMCP) Close(_ context.Context) error { return nil }
+
+// Stdio is a MCP client transport.
+//
+// Extracted from mcp-go library with changes to use in-memory in/out instead
+// of relying on starting a subcommand.
+type Stdio struct {
```

Comment:
does this work?
https://github.com/mark3labs/mcp-go/blob/91ddba5f0b9cef6fd6b89cae1009b0ab55eeb1c0/client/transport/stdio.go#L36-L48

Reply:
That's exactly what we need here 🎉